src/scheduler.c

```diff
@@ -376,14 +390,26 @@ JL_DLLEXPORT jl_task_t *jl_task_get_next(jl_value_t *trypoptask, jl_value_t *q,
             continue;
         }
 
-        jl_cpu_pause();
         jl_ptls_t ptls = ct->ptls;
         if (sleep_check_after_threshold(&start_cycles) || (ptls->tid == jl_atomic_load_relaxed(&io_loop_tid) && (!jl_atomic_load_relaxed(&_threadedregion) || wait_empty))) {
             // acquire sleep-check lock
             assert(jl_atomic_load_relaxed(&ptls->sleep_check_state) == not_sleeping);
             jl_atomic_store_relaxed(&ptls->sleep_check_state, sleeping);
             jl_fence(); // [^store_buffering_1]
             JL_PROBE_RT_SLEEP_CHECK_SLEEP(ptls);
+
+            // The place seems empty and this thread is on its way to sleeping;
+            // switch to the root task to do that. We don't do this on thread 0
+            // because we didn't start it and don't own it or its root task.
+            if (ptls->tid != 0 && ct != ptls->root_task) {
+                jl_switchto(&ptls->root_task);
+
+                // If we return here, `ct` must be runnable and someone has asked
+                // it to run, so we return it.
+                assert(jl_atomic_load_relaxed(&ct->_state) == JL_TASK_STATE_RUNNABLE);
+                return ct;
+            }
+
```

Comment:
I wonder if we can/should delay this task switch even more?

Reply:
Now that I think of it, we can refactor this slighly. The sleep atomics only make sense after we switched to the root task. And it can/should go to sleep as soon as it checks, theres no need for it to spin since we're guaranteed to have already spun before

Reply:
You're saying we should move this to the top of the `if` block, before switching `sleep_check_state` to `sleeping`?

Reply:
As discussed, I moved this code block ahead of the atomics and added a condition to the `if` to avoid sleep checking twice.